cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,261 @@ where
+            iter_ptr, tagged_ptr, 0, size, builder,
+        );
+    } else {
+        // WITH_LOOP_UNROLLING_THRESHHOLD
+
+        // === Block definitions
+        // st2g_loop_condition_block(iter_ptr)
+        let st2g_loop_condition_block = block_with_params(builder, [ValType::I64], environ)?;
+
+        // st2g_loop_body_block(iter_ptr)
+        let st2g_loop_body_block = block_with_params(builder, [ValType::I64], environ)?;
+
+        // last_stgs_block(iter_ptr)
+        let last_stgs_block = if num_required_last_stgs != 0 {
+            Some(block_with_params(builder, [ValType::I64], environ)?)
+        } else {
+            None
+        };
+
+        // next_block()
+        let next_block = block_with_params(builder, [], environ)?;
+
+        builder.ins().jump(st2g_loop_condition_block, &[iter_ptr]);
+
+        // === st2g_loop_condition_block(iter_ptr)
+        builder.switch_to_block(st2g_loop_condition_block);
+
+        let iter_ptr = builder.block_params(st2g_loop_condition_block)[0];
+
+        // TODO: the only potential optimization concern is that we now have to use a normal `icmp` over an `icmp_imm`, maybe this is slower?
+        let cond = builder
+            .ins()
+            .icmp(IntCC::UnsignedLessThan, iter_ptr, end_ptr);
+
+        match last_stgs_block {
+            Some(last_stgs_block) => {
+                canonicalise_brif(
+                    builder,
+                    cond,
+                    st2g_loop_body_block,
+                    &[iter_ptr],
+                    last_stgs_block,
+                    &[iter_ptr],
+                );
+            }
+            None => {
+                canonicalise_brif(
+                    builder,
+                    cond,
+                    st2g_loop_body_block,
+                    &[iter_ptr],
+                    next_block,
+                    &[],
+                );
+            }
+        }
+
+        // === st2g_loop_body_block(iter_ptr)
+        builder.switch_to_block(st2g_loop_body_block);
+
+        let iter_ptr = builder.block_params(st2g_loop_body_block)[0];
+
+        for i in 0..(LOOP_UNROLL_BYTES_THRESHHOLD / 32) {
+            let cur_ptr = builder.ins().iadd_imm(iter_ptr, i64::from(i) * 32);
+            builder.ins().arm64_st2g(tagged_ptr, cur_ptr);
+        }
+
+        let iter_ptr = builder
+            .ins()
+            .iadd_imm(iter_ptr, i64::from(LOOP_UNROLL_BYTES_THRESHHOLD));
+
+        builder.ins().jump(st2g_loop_condition_block, &[iter_ptr]);
+
+        builder.seal_block(st2g_loop_body_block);
+        builder.seal_block(st2g_loop_condition_block);
+
+        // === last_stgs_block(iter_ptr)
+        if let Some(last_stgs_block) = last_stgs_block {
+            builder.switch_to_block(last_stgs_block);
+
+            let iter_ptr = builder.block_params(last_stgs_block)[0];
+            let starting_offset = num_required_st2g_loops * u64::from(LOOP_UNROLL_BYTES_THRESHHOLD);
```

Comment:
This was what I wrote about it a previous comment. Clearly I could instead also just add the starting_offset directly onto `iter_ptr` instead of requiring `iter_ptr` as parameter to the basic block, but, like I said previously, we're passing the `iter_ptr` around and incrementing it in the big st2g loop anyways, so we might as well use it here. For some reason, I do still notice an `add` operation with the `starting_offset` in the generated arm machine code (when executing the tests I mentioned earlier), not sure why that is there.